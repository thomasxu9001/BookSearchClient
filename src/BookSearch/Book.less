@import './palette';

:local {
  .bookMainHeader {
    position: sticky;
    top: 0;
    background-color: @white;
  }

  .bookList {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }

  .border {
    border: 1px @black solid;
  }

  .bookCard {
    .border();
    width: 60%;
    margin: 1rem;

    &:hover {
      cursor: pointer;
      border: 1px solid @interaction;
    }

    .card__field {
      margin: 0.5rem;
    }
  }

  .searchBoxContainer {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;

    .searchInputContainer {
      display: flex;
      flex-direction: row;
      align-items: center;
      .border();
      width: 35%;
      height: 2rem;

      .searchInput {
        border: none;
        width: 100%;
        height: 95%;

        &:focus {
          outline: none;
          box-shadow: 0 0 10px #719ECE;
          border: 1px @interaction solid;
        }
      }

      .searchIcon {
        display: flex;
        align-items: center;
        justify-content: center;

        cursor: pointer;
        width: 10%;
        height: 100%;
        background-color: @interaction;
        color: @white;
      }
    }


  }

  .dropDownContainer {
    position: relative;
    width: 35rem;
    display: flex;
    justify-content: center;

    .searchHintContainer {
      position: absolute;
      background-color: @white;
      width: 35rem;
      box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);
      z-index: 1;

      .hintItem {
        padding: 0.5rem;

        &:hover {
          background: @grayLight;
        }
      }
    }
  }

  .noResult {
    margin: 1rem;
  }


  .bookContainer {
    display: flex;
    flex-direction: row;

    .bookDetailTitle {
      font-weight: bold;
      font-size: 2rem;
    }

    .bookDetailAuthor {
      margin: 1rem;
      padding-bottom: 1rem;
      border-bottom: 1px @celestialLight solid;
    }

    .bookDetailDescription {
      margin: 1rem;
      font-weight: bold;
      text-align: left;
    }

    .backToList {
      min-width: 140px;
      margin: 5px 0 0 5px;
    }
  }
}



