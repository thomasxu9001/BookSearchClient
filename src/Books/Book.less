.BookList {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.Border {
  border: 1px black solid;
}

.BookCard {
  .Border();
  width: 60%;
  margin: 1rem;

  .Card__field {
    margin: 0.5rem;
  }
}

.SearchBoxContainer {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  .Border();
}

.dropDownContainer{
  position: relative;
  width: 100%;

  .searchHintContainer {
    position: absolute;
    background-color: #f9f9f9;
    min-width: 160px;
    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
    z-index: 1;

    .hintItem {
      &:hover {
        background: #09d3ac;
      }
    }
  }
}




