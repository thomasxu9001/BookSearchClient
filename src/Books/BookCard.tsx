import React, {Component} from 'react';

interface StateProps {

}

interface State {
}


export type Props = StateProps;

export class BookCard extends Component<Props, State> {

    render() {

        return (
            <div>
                <div> Book Name</div>
                <div> Author</div>
                <div> Description</div>
            </div>
        );
    }
}
export default BookCard;



