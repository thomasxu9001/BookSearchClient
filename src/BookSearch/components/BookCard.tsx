import React, {Component} from 'react';
import {RouteComponentProps, withRouter} from 'react-router';

import {Book} from '../Book';
import s from '../Book.less';


interface StateProps {
    item: Book
}

interface State {
}


export type Props = StateProps & RouteComponentProps;

export class BookCard extends Component<Props, State> {

    handleBookCardClick = () => {
        const {item} = this.props;
        this.props.history.push('/book/' + item.id);
    };

    render() {
        const {item} = this.props;
        return (
            <div className={s.bookCard} onClick={this.handleBookCardClick}>
                <div className={s.card__field}>
                    Title: {item.title}
                </div>
                <div className={s.card__field}>
                    Author: {item.author}
                </div>
            </div>
        );
    }
}

export default withRouter(BookCard);



