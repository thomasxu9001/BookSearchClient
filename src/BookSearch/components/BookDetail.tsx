import React, {Component} from 'react';

import {Book} from '../Book';
import s from "../Book.less";

interface StateProps {
    bookId: number
}

interface State {
    error: any
    isLoaded: boolean
    book: Book | any
}


export type Props = StateProps;

export class BookDetail extends Component<Props, State> {

    constructor(props: Props) {
        super(props);
        this.state = {
            error: null,
            isLoaded: false,
            book: []
        };
    }

    componentDidMount() {
        this.getBookDetail();
    };

    getBookDetail = () => {
        const {bookId} = this.props;
        const url = "http://localhost:8000/Rest/book/id/" + bookId;

        fetch(url, {
            method: 'GET'
        })
            .then(res => res.json())
            .then(
                (result) => {
                    this.setState({
                        isLoaded: true,
                        book: result
                    });
                },
                (error) => {
                    this.setState({
                        isLoaded: true,
                        error
                    });
                }
            );
    };

    render() {
        const {book} = this.state;
        return (
            <div>
                {book &&
                <div>
                    <div className={s.bookDetailTitle}>
                        {book.title}
                    </div>
                    <div className={s.bookDetailAuthor}>
                        Author: {book.author}
                    </div>
                    <div className={s.bookDetailDescription}>
                        {book.description}
                    </div>
                </div>
                }
            </div>
        );
    }
}

export default BookDetail;



