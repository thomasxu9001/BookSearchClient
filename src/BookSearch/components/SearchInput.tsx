import React, {Component} from 'react';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {debounce} from 'throttle-debounce';

import s from '../Book.less';
import {Book} from '../Book';

interface StateProps {
    searchBook: Function
}

interface State {
    searchHint: []
    searchValue: string
}


export type Props = StateProps;

export class SearchInput extends Component<Props, State> {
    private readonly debounceSearch: debounce<any>;

    constructor(props: Props) {
        super(props);
        this.debounceSearch = debounce(300, this.searchHint.bind(this));
        this.state = {
            searchHint: [],
            searchValue: ''
        };
    };

    onChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
        const {searchBook} = this.props;

        e.persist();
        // Update input value
        this.setState({searchValue: e.target.value});
        // Update search hint logic
        if (e.target.value && e.target.value.length > 0) {
           this.debounceSearch(e.target.value);
        } else {
            this.setState({searchHint: []}, () => searchBook());
        }

    };

    searchHint = (keyword?: string) => {
        // For better practise, we should have another api which just return matched title.
        const baseUrl = "http://localhost:8000/Rest/books";
        let url = keyword ? baseUrl + '?search=' + keyword : baseUrl;
        fetch(url, {
            method: 'GET'
        })
            .then(res => res.json())
            .then(
                (result) => {
                    this.setState({
                        searchHint: result['items']
                    });
                },
                (error) => {
                }
            );
    };

    searchClickHandler = () => {
        const {searchBook} = this.props;
        const {searchValue} = this.state;

        if (searchValue) {
            this.setState({searchHint: []}, () => searchBook(searchValue));
        }
    };

    dropdownClickHandler = (title: string) => {
        const {searchBook} = this.props;
        this.setState({searchHint: [], searchValue: title}, () => searchBook(title));
    };

    renderSearchHint = () => {
        const {searchHint} = this.state;

        let items = searchHint.map((item: Book) => {
                return <div className={s.hintItem} key={item.id}
                            onClick={() => this.dropdownClickHandler(item.title)}> {item.title}</div>
            }
        );

        return <div className={s.searchHintContainer}>
            {items}
        </div>
    };

    render() {
        const {searchHint, searchValue} = this.state;

        return (

            <div className={s.searchBoxContainer}>
                <div className={s.searchInputContainer}>
                    <input className={s.searchInput} onChange={this.onChangeHandler} value={searchValue}/>
                    <div className={s.searchIcon} onClick={this.searchClickHandler}>
                        <FontAwesomeIcon icon="search"/>
                    </div>

                </div>
                <div className={s.dropDownContainer}>
                    {searchHint && searchHint.length > 0 &&
                    this.renderSearchHint()
                    }
                </div>
            </div>

        );
    }
}

export default SearchInput;



