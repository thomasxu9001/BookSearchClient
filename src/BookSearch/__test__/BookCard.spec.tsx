import React from 'react';
import {mount, ReactWrapper} from 'enzyme';
import {BookCard, Props} from '../components/BookCard';
import {Book} from "../Book";
import {createLocation, createMemoryHistory} from "history";
import {match as routerMatch} from "react-router";


const history = createMemoryHistory();
const path = `/route/:id`;

const match: routerMatch<{ id: string }> = {
    isExact: false,
    path,
    url: path.replace(':id', '1'),
    params: {id: "1"}
};
const location = createLocation(match.url);

describe('SearchInput', () => {

    let
        subject: ReactWrapper<Props>,
        item: Book
    ;

    beforeEach(() => {
        item = {
            id: 1,
            title: 'test',
            author: 'me',
            description: 'long description.'
        }
        subject = mount(
            <BookCard
                item={item}
                history={history}
                match={match}
                location={location}
            />
        );
    });

    it('looks good', async () => {
        expect(subject).toMatchSnapshot();
    });

});
