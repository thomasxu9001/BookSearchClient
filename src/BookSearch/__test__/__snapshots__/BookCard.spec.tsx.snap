// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput looks good 1`] = `ReactWrapper {}`;
